### ISSUE TITLE
Add distance functions

### ISSUE THREAD

**Issue by j-adamczyk** [scikit-fingerprints/scikit-fingerprints#228]
[RDKit distance and similarity functions](https://www.rdkit.org/docs/source/rdkit.DataStructs.cDataStructs.html) are defined for RDKit vector objects. We should implement them for NumPy arrays, with `np.array()` conversion if necessary. See also [this file](https://github.com/rdkit/rdkit/blob/0be09f8e7055879f714a802ca2590572f8036bf8/Code/DataStructs/Wrap/wrap_BitOps.cpp#L213) for metric definitions. [This paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8026075/#sec014title) also has a great list of ready formulas. [This paper](https://pubs.acs.org/doi/10.1021/ci300261r) compares metrics in chemoinformatics. [This paper](https://onlinelibrary.wiley.com/doi/10.1002/9780470027318.a9438.pub2) can also be useful, as it provides value ranges for distances typical in chemoinformatics.

Where possible, count versions should also be provided - by default, all metrics are for binary data. It must work for scikit-learn [pairwise distances function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html).

Useful reference: M. Deza, E. Deza "Dictionary of distances" (e.g. [here](https://www.uco.es/users/ma1fegan/Comunes/asignaturas/vision/Encyclopedia-of-distances-2009.pdf)). In particular, it contains formulas to transform similarity to distance (or vice versa) for some functions.

Distances to implement:
- [x] AllBit - also known as Rand similarity, Manhattan similarity
- [x] Asymmetric - also known as Simpson similarity, overlap similarity
- [x] Braun-Blanquet
- [x] CT4 (Consonni-Todeschini, variant 4) - [ref 1](https://pubs.acs.org/doi/10.1021/ci300261r), [ref 2](https://onlinelibrary.wiley.com/doi/10.1002/9780470027318.a9438.pub2), [ref 3](https://match.pmf.kg.ac.rs/electronic_versions/Match68/n2/match68n2_581-592.pdf)
- [x] Dice
- [x] Fraggle - [ref 1](https://www.rdkit.org/docs/source/rdkit.Chem.Fraggle.FraggleSim.html), [ref 2](https://rdkit.blogspot.com/2013/11/comparing-fraggle-to-other-fingerprints.html)
- [x] Harris-Lahey - [ref 1](https://pubs.acs.org/doi/10.1021/ci300261r), [ref 2](https://onlinelibrary.wiley.com/doi/10.1002/9780470027318.a9438.pub2), [ref 3](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1311335/)
- [x] Kulczynski
- [x] McConnaughey
- [x] Rogot-Goldberg
- [x] Russel
- [x] Sokal
- [x] Tanimoto

It will also be useful to implement bulk versions of those similarities, operating on whole NumPy/SciPy arrays and computing a pairwise similarity matrix. This should work similarly to `pairwise_distances` [from scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html), with required matrix `X` and optional second matrix `Y`.

Bulk distances to implement:
- [x] AllBit (Rand)
- [x] Asymmetric (Simpson)
- [x] Braun-Blanquet
- [x] CT4 (Consonni-Todeschini, variant 4)
- [x] Dice
- [ ] Fraggle
- [x] Harris-Lahey
- [x] Kulczynski
- [x] McConnaughey
- [x] Rogot-Goldberg
- [x] Russel
- [x] Sokal
- [x] Tanimoto
